net.rocketpool/image/src/app.ts

```diff
@@ -1,163 +1,262 @@
 import fs from "fs";
-import Web3 from "web3";
-import abiDecoder from "abi-decoder";
 import { Console } from "console";
 
+import { ethers } from "ethers"
+import { cleanEnv, url, str, num } from "envalid";
+
+import { RethAbi__factory } from "./contracts/factories/RethAbi__factory";
+import type { RethAbi } from "./contracts";
+
 const console = new Console(process.stderr);
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
-interface Event {
+interface EventRow {
   eventName: "TokensMinted" | "TokensBurned";
   amount: string;
   ethAmount: string;
   blockNumber: number;
   blockHash: string;
-  blockTime: number;
+  eventTime: number;
   transactionIndex: number;
   transactionHash: string;
   logIndex: number;
 }
 
+interface OdfReadRowsResult {
+  etag: number;
+  hasMore: boolean;
+}
+
+type Provider = ethers.providers.JsonRpcProvider | ethers.providers.WebSocketProvider;
+type OnEventCallback = (_: EventRow) => boolean;
+type AllBlockRangeHasProcessed = boolean;
+
 /////////////////////////////////////////////////////////////////////////////////////////
 
-async function readEventsFromETH(
-  startBlock,
-  endBlock,
-  clb: (_: Event) => boolean,
-  stride = 10000
-) {
-  console.log(
-    "Considering events in block range [%s, %s] (%s blocks)",
+async function main() {
+  const env = cleanEnv(process.env, {
+    // Application:
+    ETH_NODE_PROVIDER_URL: url(),
+    BLOCK_BATCH_SIZE: num(),
+    // ODF interface:
+    ODF_BATCH_SIZE: num(),
+    ODF_NEW_ETAG_PATH: str(),
+    ODF_NEW_HAS_MORE_DATA_PATH: str(),
+  })
+
+  // See: https://etherscan.io/token/0xae78736cd615f374d3085123a210448e74fc6393
+  //      https://etherscan.io/address/0xae78736cd615f374d3085123a210448e74fc6393
+  const rethAddress = "0xae78736cd615f374d3085123a210448e74fc6393";
+  const rethStartBlock = 13325304;
+
+  // Init
+  const providerUrl = env.ETH_NODE_PROVIDER_URL;
+
+  console.log(`ETH node provider URL: ${providerUrl}`);
+
+  const provider = getProvider(providerUrl)
+  const rethContract = RethAbi__factory.connect(rethAddress, provider)
+
+  // Block range
+  const lastSeenBlock = process.env.ODF_ETAG ? Number(process.env.ODF_ETAG) : 0;
```

Comment:
Can we move it to env object declared above?
And I'm not a huge fan of ternary operators, cause they often are hard to read so here I can propose to use next `const lastSeenBlock = parseInti(process.env.ODF_ETAG || 0);`

Reply:
can this be moved into `cleanEnv` above or at least next to it so all env vars were read in one place?

Reply:
> Can we move it to env object declared above?

Yes, for sure -- previously I missed the default value configuration in `envalid`

Fixed